File: is_production/production/report/rental_equipment_hours/rental_equipment_hours.js
/* eslint-disable */

frappe.query_reports["Rental Equipment Hours"] = {
    filters: [
        {
            fieldname: "start_date",
            label: __("Start Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.add_days(frappe.datetime.get_today(), -7)
        },
        {
            fieldname: "end_date",
            label: __("End Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.get_today()
        },
        {
            fieldname: "site",
            label: __("Site"),
            fieldtype: "Link",
            options: "Location",
            reqd: 1
        },
        {
            fieldname: "shift",
            label: __("Shift"),
            fieldtype: "Select",
            options: ["", "Day", "Night"],
            reqd: 0
        }
    ],

    formatter: function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "working_hours" && data && data.working_hours > 100) {
            value = "<span style='color:green; font-weight:bold'>" + value + "</span>";
        }

        if (data && data.asset_name === "🔢 Grand Total") {
            value = "<span style='font-weight:bold'>" + value + "</span>";
        }

        return value;
    }
};

File: is_production/production/report/rental_equipment_hours/rental_equipment_hours.py
import frappe
from frappe.utils import flt

def execute(filters=None):
    if not filters:
        return [], []

    # --- Report Columns ---
    columns = [
        {"label": "Asset", "fieldname": "asset_name", "fieldtype": "Data", "width": 220},
        {"label": "Category", "fieldname": "asset_category", "fieldtype": "Data", "width": 150},
        {"label": "Working Hours", "fieldname": "working_hours", "fieldtype": "Float", "width": 140},
    ]

    # --- Query Conditions ---
    conditions = [
        "ph.shift_date BETWEEN %(start_date)s AND %(end_date)s",
        "ph.location = %(site)s",
        "pa.asset_category NOT IN ('Dozer','ADT','Excavator')"  # 🚫 exclude these
    ]
    if filters.get("shift"):
        conditions.append("ph.shift = %(shift)s")

    where_clause = " AND ".join(conditions)

    # --- Fetch cumulative working hours ---
    data = frappe.db.sql(
        f"""
        SELECT
            pa.asset_name,
            pa.asset_category,
            SUM(pa.working_hours) as working_hours
        FROM `tabPre-Use Hours` ph
        JOIN `tabPre-use Assets` pa ON pa.parent = ph.name
        WHERE {where_clause}
        GROUP BY pa.asset_name, pa.asset_category
        ORDER BY pa.asset_category, pa.asset_name
        """,
        filters,
        as_dict=True
    )

    # --- Add Grand Total Row ---
    if data:
        total_hours = sum(flt(d.working_hours) for d in data)
        data.append({
            "asset_name": "🔢 Grand Total",
            "asset_category": "",
            "working_hours": total_hours
        })

    return columns, data



